File: scripts/ticket-mcp/go/internal/runner/run_ticket.go
package runner

import (
	"bytes"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

const maxBufferSize = 10 * 1024 * 1024

func findRepoRoot() string {
	if env := os.Getenv("TICKET_MCP_REPO_ROOT"); env != "" {
		return env
	}
	exe, err := os.Executable()
	if err != nil {
		return ""
	}
	exe, err = filepath.EvalSymlinks(exe)
	if err != nil {
		exe = os.Args[0]
	}
	dir := filepath.Dir(exe)
	for {
		if _, err := os.Stat(filepath.Join(dir, ".git")); err == nil {
			if _, err := os.Stat(filepath.Join(dir, "scripts", "ticket.sh")); err == nil {
				return dir
			}
		}
		parent := filepath.Dir(dir)
		if parent == dir {
			return ""
		}
		dir = parent
	}
}

var repoRoot = findRepoRoot()

func ticketShPath() string {
	if env := os.Getenv("TICKET_SH"); env != "" {
		cleaned := filepath.Clean(env)
		root := repoRoot
		if root != "" && strings.HasPrefix(cleaned, root) {
			return cleaned
		}
	}
	return filepath.Join(repoRoot, "scripts", "ticket.sh")
}

func RunTicket(args []string, extraEnv map[string]string) (string, error) {
	ticketSh := ticketShPath()
	cleaned := filepath.Clean(ticketSh)
	if !strings.HasPrefix(cleaned, repoRoot) {
		return "", fmt.Errorf("ticket.sh path %s is outside repo root %s", cleaned, repoRoot)
	}
	cmd := exec.Command("bash", append([]string{ticketSh}, args...)...)
	env := os.Environ()
	for k, v := range extraEnv {
		env = append(env, k+"="+v)
	}
	cmd.Env = env
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr
	if err := cmd.Run(); err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			trimmed := strings.TrimSpace(stderr.String())
			if trimmed == "" {
				trimmed = err.Error()
			}
			return "", fmt.Errorf("ticket.sh failed (exit code %d): %s", exitErr.ExitCode(), trimmed)
		}
		return "", fmt.Errorf("ticket.sh failed: %s", err.Error())
	}
	if stdout.Len() > maxBufferSize {
		return stdout.String()[:maxBufferSize], nil
	}
	return stdout.String(), nil
}

func RepoRoot() string {
	return repoRoot
}

File: scripts/ticket-mcp/lib/run-ticket.js
import { execFile } from 'node:child_process';
import { promisify } from 'node:util';
import { fileURLToPath } from 'node:url';
import path from 'node:path';

const execFileAsync = promisify(execFile);

const REPO_ROOT = path.resolve(fileURLToPath(import.meta.url), '../../../..');
const DEFAULT_TICKET_SH = path.join(REPO_ROOT, 'scripts', 'ticket.sh');

export async function runTicket(args, extraEnv = {}) {
  const ticketSh = extraEnv.TICKET_SH ?? DEFAULT_TICKET_SH;
  const env = { ...process.env, ...extraEnv };
  delete env.TICKET_SH;

  try {
    const { stdout } = await execFileAsync('bash', [ticketSh, ...args], {
      env,
      maxBuffer: 10 * 1024 * 1024,
    });
    return stdout;
  } catch (err) {
    const msg = err.stderr?.trim() || err.message;
    throw new Error(`ticket.sh failed (exit code ${err.code}): ${msg}`);
  }
}
